msal/src/main/java/com/microsoft/identity/nativeauth/statemachine/states/MFAStates.kt

```diff
@@ -0,0 +1,220 @@
+            return AwaitingMFAState(parcel)
+        }
+
+        override fun newArray(size: Int): Array<AwaitingMFAState?> {
+            return arrayOfNulls(size)
+        }
+    }
+}
+
+class MFARequiredState(
+    override val continuationToken: String,
+    override val correlationId: String,
+    private val scopes: List<String>?,
+    private val config: NativeAuthPublicClientApplicationConfiguration
+) : BaseState(continuationToken = continuationToken, correlationId = correlationId), State, Parcelable {
+    private val TAG: String = MFARequiredState::class.java.simpleName
+
+    // Challenge specified auth methods
+    suspend fun sendChallenge(authMethodId: String): MFARequiredResult {
+        return withContext(Dispatchers.IO) {
+            // if /challenge returns HTTP 200
+            if (true) {
+                MFARequiredResult.VerificationRequired(
+                    nextState = MFARequiredState(
+                        continuationToken = continuationToken,
+                        correlationId = correlationId,
+                        scopes = scopes,
+                        config = config
+                    ),
+                    codeLength = 6,
+                    sentTo = "user@contoso.com",
+                    channel = "email"
+                )
+            } else {
+                // /challenge returns introspect_required
+                // call /introspect and return authMethods
+                MFARequiredResult.SelectionRequired(
+                    nextState = MFARequiredState(
+                        continuationToken = continuationToken,
+                        correlationId = correlationId,
+                        scopes = scopes,
+                        config = config
+                    ),
+                    authMethods = listOf(AuthMethod(
+                        id = "1234",
+                        challengeType = "oob",
+                        challengeChannel = "email",
+                        loginHint = "u***@**.com"
+                    ))
+                )
+            }
+        }
+    }
+
+    // Call /introspect
+    suspend fun getAuthMethods(): MFAGetAuthMethodsResult {
+        return MFARequiredResult.SelectionRequired(
+            nextState = MFARequiredState(
+                continuationToken = continuationToken,
+                correlationId = correlationId,
+                scopes = scopes,
+                config = config
+            ),
+            authMethods = listOf(AuthMethod(
+                id = "1234",
+                challengeType = "oob",
+                challengeChannel = "email",
+                loginHint = "u***@**.com"
+            ))
+        )
+    }
+
+    // Call /token
+    suspend fun submitChallenge(code: Int): MFASubmitChallengeResult {
+        // If /token returns HTTP 200
+        return if (true) {
+            SignInResult.DummyComplete()
+        } else {
+            // If /token returns another mfa_required error
```

Comment:
Does this means that the `MFASubmitChallengeResult` has `SignInComplete` and  `MFARequired` type? I thought we wanted to limit to only a second factor authentication.

Reply:
Correct. We're currently discussing this in the design doc.

Reply:
I've removed it for now, and will put it back after we conclude the discussion with Kaushik.